Support SDXL and its distributed inference

Zars19 opened the pull request:
The idea of patch parallelism comes from the CVPR 2024 paper [Distrifusion](https://github.com/mit-han-lab/distrifuser). In order to reduce the difficulty of implementation, all communications in the example are synchronous.

This can help SDXL achieve better performance, especially when the resolution is very high

A100, 50 steps, 2048x2048, SDXL
Framework | sync_mode | n_gpu | latency(s) | speed_up | memory(MiB)
-- | -- | -- | -- | -- | --
Torch | - | 1 | 25.25 | 1x | 42147
TRT | - | 1 | 21.98 | 1.15x | 42895
DistrFusion(Torch) | split_batch | 2 | 13.33 | 1.89x | 40173
Ours | split_batch | 2 | 11.69 | 2.16x | 42675
DistrFusion(Torch) | corrected_async_gn | 4 | 8.27 | 3.05x | 49087
DistrFusion(Torch) | full_sync | 4 | 8.64 | 2.92x | 51943
Ours | full_sync | 4 | 7.73 | 3.27x | 43073

juney-nvidia commented:
@Zars19 thanks for the contribution to TensorRT-LLM!

@nv-guomingz  can you help take care of this? :) 

Thanks
June

nv-guomingz commented:
> @Zars19 thanks for the contribution to TensorRT-LLM!
> 
> @nv-guomingz can you help take care of this? :)
> 
> Thanks June

Sure, I'll collobrate with @Zars19 for enabling SDXL with TRT-LLM.

nv-guomingz commented:
Hi @Zars19 , could u please resolve the code conflicts firstly?

Zars19 commented:
> Hi @Zars19 , could u please resolve the code conflicts firstly?

I have resolved the conflict :)

Zars19 commented:
@nv-guomingz OK, I have completed the git rebase

DefTruth commented on tensorrt_llm/builder.py:
```diff
@@ -178,16 +179,15 @@ def create_builder_config(self,
             if fp8:
                 config.set_flag(trt.BuilderFlag.FP8)
-                config.set_flag(trt.BuilderFlag.OBEY_PRECISION_CONSTRAINTS)
+
```
这里的修改覆盖了最新的builder.py的逻辑，会导致fp8 engine在trtllm-build阶段编译错误，需要使用最新的逻辑。

DefTruth replied:
@Zars19 @nv-guomingz

Zars19 replied:
你说得对，我重新提交了代码修复这个问题

lmxyy commented:
Any updates on the codereview?

Zars19 commented:
> Any updates on the code review?

After rebasing the code, I haven't received feedback for a while now

hchings commented:
Hi @Zars19, thanks for your patience. We recently resumed reviewing this with the latest TRT-LLM and noticed some issues so far:

1. `dtype` passed into `get_timestep_embedding`([line](https://github.com/NVIDIA/TensorRT-LLM/pull/1514/files#diff-fb9a6774ec2fb22f069f3870d6185cea5c58ac107b51a23ac17f52a85fa0a01aR29)) but time embedding is still fixed to fp32.  -- this is a minor fix and we've added it.
2. Image output when `n_gpu=2` is incorrect. It seems that the `DistriUNetPP` or maybe the scatter/sparse op has issues. Can you help taking a look? Below are output imgs of this MR on [Distrifusion's default prompt](https://github.com/mit-han-lab/distrifuser/blob/main/scripts/run_sdxl.py#L25), you can see the part dealing w/ activation patch is not right. (Config is the same as this MR's, except that randomseed is set to 1234 for easier comparison w/ Distrifuion original repo).

<img width="557" alt="image" src="https://github.com/user-attachments/assets/3f5d755f-4f72-4521-9934-679a578a65e9">

<br>

3. `n_gpu=4` with either TRT-LLM v0.13 or our internal latest main branch failed to build with the following error:
<img width="1352" alt="image" src="https://github.com/user-attachments/assets/d98e2eb6-0f5a-402c-b763-a6c38b70fe17">
<br>
<br>
A few other questions:

- In your benchmark, do the rows that marked as `Ours / split_batch` and `Ours / full_async` both use batch splitting? I don't see there's a flag in script to enable/disable batch splitting in your scripts. 

- Could you share how did you measure the peak mem usage? I got similar speedup as yours (other than I cannot run 4 gpus yet), but I observed much less peak mem usuage of DistrFusion(Torch).



<br class="Apple-interchange-newline">




<br class="Apple-interchange-newline">

Zars19 commented:
@hchings Thank you for your feedback! I've rebased with the latest official code, debugged, and updated some issues.

1. I've fixed the fp32 constant type mismatch issue in the latest commit.
2. I couldn't reproduce the image generation error with the default prompt. Using the following commands, the generated image is consistent with the original result:
   ```
   mpirun -n 2 python build_sdxl_unet.py --size 1024
   mpirun -n 2 python run_sdxl.py --size 1024 --prompt "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k" --seed 1234
   ```
![image](https://github.com/user-attachments/assets/38500724-6719-4739-aa1c-98052b6f8877)

3. In the new commit, I manually fused pre-padding or post-padding into conv2d, resolving the compilation issue.

Responses to other issues:
* Yes, splitting the batch enhances the efficiency of distributed inference, hence I prioritized processing the two classifier-guidance-free batches separately. This is done by default without a user option.
* The table shows performance and memory usage for generating images at a resolution of 2048x2048, as optimizations yield a higher speedup ratio at higher resolutions. Generating 1024x1024 images requires less than half of the memory. Could you please confirm if your comparison is based on the memory data at the 1024 resolution size?

hchings commented:
Hi @Zars19 , thanks for the fixes!

- Re # 2.  I pulled your latest changes and can still reproduce it with `--size=2048`. Attached another result below. However, it seems that it's a limitation of SDXL rather than distributed implementation issue, as I can repro similar broken images w/ the [Distrifuser repo](https://github.com/mit-han-lab/distrifuser/tree/main) and worse img with vanilla HuggingFace diffusers. If this is true, then it seems there's not much need to test speedup for 2048x2048 and 3840x3840 for this MR as did in the paper, as the output img won't be practical. Pls correct me if I miss sth.

  ```
  # this MR
  mpirun -n 4 python build_sdxl_unet.py --size 2048
  mpirun -n 4 python run_sdxl.py --size 2048 --prompt "flowers, rabbit" --seed 1234
  
  # Distrifuser
  torchrun --nproc_per_node=4 scripts/run_sdxl.py --image_size 2048 --prompt "flowers, rabbit" --output_path xxx
  ```

  <img width="932" alt="image" src="https://github.com/user-attachments/assets/3c867b4b-f05f-45b8-a46f-3a8d19d0d4a9">


</br>

- The latest commit of this MR introduced a shape mismatch issue at ResNet `forward()` ([line](https://github.com/NVIDIA/TensorRT-LLM/pull/1514/files#diff-777324a262e374a3c32aff14e5602d0ffee55ae1b438e47d3d7a79b4b313e4a0R252)), where `input_tensors` should be in the shape of `(1, 320, 128, 128)` but is now `(1, 320, 126, 126)`.  Can you double check your new modification in `conv2d.py`?
  
  Repro (after build):
  ```
  mpirun -n 2 python build_sdxl_unet.py --size 1024
  ```
  
  <img width="1076" alt="image" src="https://github.com/user-attachments/assets/44c07eb0-36e3-4dbd-a85e-0248aa24166a">

Zars19 commented:
@hchings Thanks for your reply.

* Yes, the 2048 resolution images generated by distrifuser and this PR are consistent with the original output of the SDXL model. I believe the paper's comparison of speedup at higher resolutions aims to demonstrate a general trend: that this distributed method benefits diffusion models with improved performance at higher resolutions. However, for the SDXL model specifically, generating resolutions above the trained 1024 often results in pattern repetition and artifacts, leading to invalid outputs.
```python
import torch
from diffusers import StableDiffusionXLPipeline

pipeline = StableDiffusionXLPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0",
    torch_dtype=torch.float16,
    use_safetensors=True,
)
pipeline.to('cuda')

seed = 1234
size = 2048
#prompt = "flowers, rabbit"
prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"

image = pipeline(
    prompt=prompt,
    generator=torch.Generator(device="cuda").manual_seed(seed),
    height=size, width=size).images[0]

image.save(f"output.png")
```
Picture generated by the original SDXL:
<img src="https://github.com/user-attachments/assets/1a773158-b4f9-430c-b283-fa4ed5ed1f42" width="500">

* Previously, I set the default values of the `pre_padding` and `post_padding` parameters in the `conv2d` function of `functional.py` to (0,0). These had higher priority and were causing shape issues by making the regular padding parameter ineffective. This issue has now been resolved.

hchings commented:
Hi @Zars19, FYI that we're doing some final wrap-ups of this and will merge it soon. Thanks!

hchings commented:
Hi @Zars19, we've merged this internally and it will show up in the upcoming public release under `/example/sdxl`.
Thanks you for the contribution & patience.
